components/script/dom/imagedata.rs

```diff
@@ -47,122 +51,92 @@ impl ImageData {
+                        .map_err(|_| Error::JSFailed)?);
 
-        let height = len / width;
-        if opt_height.is_some_and(|x| height != x) {
-            return Err(Error::IndexSize);
+            Self::Constructor_(global, None, can_gc, data, width, Some(height), &settings)
+        } else {
+            Self::Constructor(global, None, can_gc, width, height, &settings)
         }
-
-        let imagedata = Box::new(ImageData {
-            reflector_: Reflector::new(),
-            width,
-            height,
-            data: heap_typed_array,
-        });
-
-        Ok(reflect_dom_object_with_proto(
-            imagedata, global, proto, can_gc,
-        ))
     }
 
-    fn new_without_data(
+    #[allow(clippy::too_many_arguments)]
+    /// <https://html.spec.whatwg.org/multipage/#initialize-an-imagedata-object>
+    fn initialize(
+        pixels_per_row: u32,
+        rows: u32,
+        settings: &ImageDataSettings,
+        source: Option<CustomAutoRooterGuard<Uint8ClampedArray>>,
+        default_color_space: Option<PredefinedColorSpace>,
         global: &GlobalScope,
         proto: Option<HandleObject>,
-        width: u32,
-        height: u32,
         can_gc: CanGc,
     ) -> Fallible<DomRoot<ImageData>> {
-        // If one or both of sw and sh are zero, then throw an "IndexSizeError" DOMException.
-        if width == 0 || height == 0 {
-            return Err(Error::IndexSize);
-        }
-
-        // When a constructor is called for an ImageData that is too large, other browsers throw
-        // IndexSizeError rather than RangeError here, so we do the same.
-        let len =
-            pixels::compute_rgba8_byte_length_if_within_limit(width as usize, height as usize)
-                .ok_or(Error::IndexSize)?;
-
-        let cx = GlobalScope::get_cx();
-
-        let heap_typed_array =
-            create_heap_buffer_source_with_length::<ClampedU8>(cx, len as u32, can_gc)?;
-
-        let imagedata = Box::new(ImageData {
-            reflector_: Reflector::new(),
-            width,
-            height,
-            data: heap_typed_array,
-        });
+        // 1. If source was given:
+        let data = if let Some(source) = source {
+            HeapBufferSource::<ClampedU8>::from_view(source)
+        } else {
+            // 2. Otherwise (source was not given):
+            match settings.pixelFormat {
+                ImageDataPixelFormat::Rgba_unorm8 => {
+                    // 1. If settings["pixelFormat"] is "rgba-unorm8",
+                    // then initialize the data attribute of imageData to a new Uint8ClampedArray object.
+                    // The Uint8ClampedArray object must use a new ArrayBuffer for its storage,
+                    // and must have a zero byte offset and byte length equal to the length of its storage, in bytes.
+                    // The storage ArrayBuffer must have a length of 4 × rows × pixelsPerRow bytes.
+                    create_heap_buffer_source_with_length(
```

Comment:
Add a note for "[rgba-float16](https://html.spec.whatwg.org/multipage/imagebitmap-and-animations.html#dom-imagedatapixelformat-rgba-unorm8)"?

Also, you could add "3. If the storage [ArrayBuffer](https://webidl.spec.whatwg.org/#idl-ArrayBuffer) could not be allocated, then rethrow the [RangeError](https://tc39.es/ecma262/#sec-native-error-types-used-in-this-standard-rangeerror) thrown by JavaScript, and return.", and add a note that is  done with the result of `create_heap_buffer_source_with_length`(is it?).

Reply:
It is.